### PR TITLE
feat(aws-codebuild): add ProjectFileSystemLocation property to codebuild

### LINKED ISSUE #6533: Support for CodeBuild builds on AWS EFS mounts
Add support for CodeBuild to build on AWS EFS file system mounts using the new *ProjectFileSystemLocation* .
more info here https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-codebuild-project.html

### Use Case

Create AWS CodeBuild builds on Amazon EFS.

### Proposed Solution

as per the documentation https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-codebuild-project.html

### Other

* [x] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by nataibi**
fixes #6533 

- [x] Testing
Unit test added (prefer not to modify an existing test, otherwise, it's probably a breaking change)
Integration tests added
- [x] Docs
jsdocs: All public APIs documented
- [x] Title: add ProjectFileSystemLocations property to Codebuild

### CODE REVIEW

File: packages/@aws-cdk/aws-codebuild/README.md [hunk 1]
```diff
@@ -379,3 +379,33 @@ const project = new codebuild.Project(this, 'MyProject', {
 
 project.connections.allowTo(loadBalancer, ec2.Port.tcp(443));
 ```
+
+## Project File System Location EFS
+
+Add support for CodeBuild to build on AWS EFS file system mounts using
+the new ProjectFileSystemLocation .
+The `fileSystemLocations` property which accepts a list `ProjectFileSystemLocation`
+as represented by the interface `IFileSystemLocations`
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codebuild/README.md]: Missing period at the end of this sentence.

File: packages/@aws-cdk/aws-codebuild/README.md [hunk 2]
```diff
@@ -379,3 +379,33 @@ const project = new codebuild.Project(this, 'MyProject', {
 
 project.connections.allowTo(loadBalancer, ec2.Port.tcp(443));
 ```
+
+## Project File System Location EFS
+
+Add support for CodeBuild to build on AWS EFS file system mounts using
+the new ProjectFileSystemLocation .
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codebuild/README.md]: Unnecessary space before the `.`.

File: packages/@aws-cdk/aws-codebuild/README.md [hunk 3]
```diff
@@ -379,3 +379,33 @@ const project = new codebuild.Project(this, 'MyProject', {
...
+The `fileSystemLocations` property which accepts a list `ProjectFileSystemLocation`
+as represented by the interface `IFileSystemLocations`
+The only supported file system type is `EFS`
+
+For Example:
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codebuild/README.md]:
```suggestion
For example:
```

File: packages/@aws-cdk/aws-codebuild/README.md [hunk 4]
```diff
@@ -379,3 +379,33 @@ const project = new codebuild.Project(this, 'MyProject', {
...
 project.connections.allowTo(loadBalancer, ec2.Port.tcp(443));
 ```
+
+## Project File System Location EFS
+
+Add support for CodeBuild to build on AWS EFS file system mounts using
+the new ProjectFileSystemLocation .
+The `fileSystemLocations` property which accepts a list `ProjectFileSystemLocation`
+as represented by the interface `IFileSystemLocations`
+The only supported file system type is `EFS`
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codebuild/README.md]: Same (missing period at the end of this sentence).

File: packages/@aws-cdk/aws-codebuild/lib/project.ts [hunk 5]
```diff
@@ -770,6 +787,18 @@ export class Project extends ProjectBase {
...
   }
 
+  /**
+   * Adds a fileSystemLocation to the Project.
+   *
+   * @param fileSystemLocation the fileSystemLocation to add
+   */
+  public addFileSystemLocations(fileSystemLocation: IFileSystemLocation): void {
+    const fileSystemConfig = fileSystemLocation.bind(this, this);
+    if (fileSystemConfig) {
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codebuild/lib/project.ts]: Why this `if`? `IFileSystemLocation.bind` never returns `undefined`....

File: packages/@aws-cdk/aws-codebuild/README.md [hunk 6]
```diff
@@ -379,3 +379,33 @@ const project = new codebuild.Project(this, 'MyProject', {
...
+    }),
+    fileSystemLocations: [{
+        identifier: "myidentifier",
+        location: "fs-c8d04839.efs.eu-west-2.amazonaws.com:/mnt",
+        mountOptions: "nfsvers=4.1,rsize=1048576,wsize=1048576,hard,timeo=600,retrans=2",
+        mountPoint: "/media",
+        type: "EFS"
+
+    }]
+});
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codebuild/README.md]: This example is now out of date.

File: packages/@aws-cdk/aws-codebuild/lib/project.ts [hunk 7]
```diff
@@ -1515,3 +1550,52 @@ export enum BuildEnvironmentVariableType {
    */
   SECRETS_MANAGER = 'SECRETS_MANAGER'
 }
+
+export interface FileSystemConfig {
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codebuild/lib/project.ts]: Can you move all of this, starting from this line, to a separate file?

File: packages/@aws-cdk/aws-codebuild/lib/project.ts [hunk 8]
```diff
@@ -1515,3 +1550,52 @@ export enum BuildEnvironmentVariableType {
...
+        mountPoint: this.props.mountPoint,
+        type: 'EFS',
+      },
+    };
+  }
+}
+/**
+ * The abstract interface of a CodeBuild FileSystemLocation.
+ */
+export interface IGenericFileSystemLocationProps {
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codebuild/lib/project.ts]: This name is wrong. It's not a generic interface; quite the opposite, these are properties specific to the EFS FileSystemLocation.

So, the name should be `EfsFileSystemLocationProps`.

File: packages/@aws-cdk/aws-codebuild/lib/project.ts [hunk 9]
```diff
@@ -1515,3 +1550,52 @@ export enum BuildEnvironmentVariableType {
...
+  readonly identifier: string;
+  readonly location: string;
+  readonly mountOptions?: string;
+  readonly mountPoint: string;
+  readonly type: string;
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codebuild/lib/project.ts]: `type` should not be here. It's always `EFS` in this case (the property is also never used).

File: packages/@aws-cdk/aws-codebuild/package.json [hunk 10]
```diff
@@ -208,7 +208,18 @@
...
+      "docs-public-apis:@aws-cdk/aws-codebuild.IGenericFileSystemLocationProps.location",
+      "docs-public-apis:@aws-cdk/aws-codebuild.IGenericFileSystemLocationProps.mountPoint",
+      "docs-public-apis:@aws-cdk/aws-codebuild.IGenericFileSystemLocationProps.type",
+      "docs-public-apis:@aws-cdk/aws-codebuild.IGenericFileSystemLocationProps.mountOptions",
+      "docs-public-apis:@aws-cdk/aws-codebuild.IFileSystemLocation",
+      "docs-public-apis:@aws-cdk/aws-codebuild.IFileSystemLocation.bind",
+      "docs-public-apis:@aws-cdk/aws-codebuild.FileSystemConfig",
+      "docs-public-apis:@aws-cdk/aws-codebuild.FileSystemConfig.location",
+      "docs-public-apis:@aws-cdk/aws-codebuild.FileSystemLocation",
+      "docs-public-apis:@aws-cdk/aws-codebuild.FileSystemLocation.efs"
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codebuild/package.json]: Can you write those docs instead? :)

File: packages/@aws-cdk/aws-codebuild/lib/project.ts (hunk 8 above)
Comment (nataibi) [reply packages/@aws-cdk/aws-codebuild/lib/project.ts]: Makes sense, thanks for spotting this

File: packages/@aws-cdk/aws-codebuild/lib/project.ts [hunk 11]
```diff
@@ -770,6 +788,16 @@ export class Project extends ProjectBase {
     }
   }
 
+  /**
+   * Adds a fileSystemLocation to the Project.
+   *
+   * @param fileSystemLocation the fileSystemLocation to add
+   */
+  public addFileSystemLocations(fileSystemLocation: IFileSystemLocation): void {
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codebuild/lib/project.ts]: Oh, one more thing - rename this method to `addFileSystemLocation` (no `s` at the end).

File: packages/@aws-cdk/aws-codebuild/lib/project.ts [hunk 12]
```diff
@@ -508,6 +509,16 @@ export interface CommonProjectProps {
...
    */
   readonly allowAllOutbound?: boolean;
+
+  /**
+   * An  ProjectFileSystemLocation objects for a CodeBuild build project.
+   *
+   * A ProjectFileSystemLocation object specifies the identifier, location, mountOptions, mountPoint,
+   * and type of a file system created using Amazon Elastic File System.
+   *
+   * @default false
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codebuild/lib/project.ts]:
```suggestion
   * @default - no file system locations
```

**Comment by nataibi** [comment]
> This is awesome @nataibi !
> 
> Just one more thing - we use 2 spaces for indentation in our TypeScript files. Your new files use 4 spaces. If you could just fix that one small detail, I'll merge this in!

thank you for your support, appreciated

**Comment by logemann** [comment]
and thanks for the well done examples and the README. Will give it a try now...

**Pull request merged**